### PR TITLE
StdlibRef: Sequence.withContiguousStorageIfAvailable(_:) abstract needs a rewrite

### PR DESCRIPTION
**Pull request by invalidname**
Update source documentation comments for `withContiguousStorageIfAvailable(_:)` in `Sequence` and `MutableCollection` to improve style (single sentence summary), include parameter and return value blocks, and include details from the original swift-evolution proposal.

### CODE REVIEW

File: stdlib/public/core/MutableCollection.swift [hunk 1]
```diff
@@ -182,16 +182,32 @@ where SubSequence: MutableCollection
...
+  ///
+  /// This method makes no guarantees about the state of the collection if
+  /// the closure throws an error. Any changes made before the throw may
+  /// or may not be reflected in the collection. This is because the closure
+  /// could receive direct access to the collection's internal storage, or a
+  /// temporary copy. Therefore, the closure should always perform any
+  /// necessary cleanup.
+  ///
+  /// - Parameter body: A closure that receives an in-out
+  /// `UnsafeMutableBufferPointer` to the collection's contiguous storage.
```
Comment (glessard) [inline stdlib/public/core/MutableCollection.swift:207]: The UMBP is `inout` in order to enable ready use of `MutableCollection` API that require `self` to be mutable. In all implementations in the stdlib, swapping out the UMBP is an error that leads to a crash. Given that, mentioning the in-out-ness of the argument simply shines a light on a confusing aspect without a necessary warning.

**Review by glessard — COMMENTED** [review]
Clearly an improvement!

File: stdlib/public/core/MutableCollection.swift (hunk 1 above)
Comment (invalidname) [reply stdlib/public/core/MutableCollection.swift:207]: What would you think of adding another sentence of guidance to that effect, something like:

> Use the buffer pointer directly; do not replace it, as this leads to a crash in all implementations of this method within the standard library.

Comment (Catfish-Man) [reply stdlib/public/core/MutableCollection.swift:207]: I actually just ran into this myself and had to be corrected by my team, so mentioning that does seem good!

Comment (invalidname) [reply stdlib/public/core/MutableCollection.swift]: Added a line about not replacing the UMBP.

**Review by glessard — APPROVED** [review]
Thanks!

File: stdlib/public/core/Sequence.swift [hunk 2]
```diff
@@ -384,16 +384,25 @@ public protocol Sequence {
...
+  /// This method calls `body(p)`, where `p` is a pointer to the collection’s
+  /// contiguous storage. If no such storage exists, the collection creates it.
+  /// If the collection doesn’t support an internal representation in a form of
+  /// contiguous storage, the method doesn’t call `body` and returns `nil`.
   ///
   /// A `Collection` that provides its own implementation of this method
-  /// must also guarantee that an equivalent buffer of its `SubSequence` 
-  /// can be generated by advancing the pointer by the distance to the
-  /// slice's `startIndex`.
+  /// must also guarantee that advancing the pointer by the distance to the
```
Comment (xwu) [inline stdlib/public/core/Sequence.swift:395]:
```suggestion
  /// must also guarantee that advancing the pointer by the distance to a
```

It's unclear what would be referred to here by "_the_ slice." Overall, I feel the original wording was more easily comprehended here.

**Review by xwu — COMMENTED** [review]
This reads very nicely overall! One suggestion below and one nit--seems there's a mix of smart apostrophes and straight apostrophes; can you make them all the latter?

**Comment by glessard** [comment]
@swift-ci please smoke test

**Comment by glessard** [comment]
@invalidname I completely forgot this existed, and discussions among the stdlib team led us to write some clarifications. We should figure out how to merge without removing information.

**Comment by invalidname** [comment]
> @invalidname I completely forgot this existed, and discussions among the stdlib team led us to write some clarifications. We should figure out how to merge without removing information.

Sorry for the delay; it spent a pretty long time in the editing queue, so it was easily overlooked. Maybe email me with how to proceed?

**Comment by invalidname** [comment]
Now that #40027 has merged, this PR is going to have hopeless conflicts that can't realistically be resolved. I'm going to create a new branch off of main to pick up those changes, then apply the rewrites and new material from this PR.

**Comment by invalidname** [comment]
Abandoning this PR in favor of #40334, which incorporates the changes that #40027 merged into main.

**Pull request closed without merging**